File: Trees/9. Binary Tree Right Side View/solution.exs
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec right_side_view(root :: TreeNode.t | nil) :: [integer]
  def right_side_view(root) do
    solve(root, 0, %{})
    |> Enum.sort()
    |> Enum.map(fn {_, k} -> k end)
  end

  def solve(nil, level, record), do: record

  def solve(%{left: left, right: right, val: val}, level, record) do
      record = Map.put(record, level, val)
      acc = solve(left, level+1,  record)
      solve(right, level+1, acc)
  end
end

File: Trees/9. Binary Tree Right Side View/solution.scala
/** Definition for a binary tree node. class TreeNode(_value: Int = 0, _left:
  * TreeNode = null, _right: TreeNode = null) { var value: Int = _value var
  * left: TreeNode = _left var right: TreeNode = _right }
  */
object Solution {
  def rightSideView(root: TreeNode): List[Int] = {
    if (root == null) return List()
    var nodes: List[TreeNode] = List(root)
    var result: List[Int] = List()
    while (!nodes.isEmpty) {
      result = result :+ nodes.last.value
      nodes =
        nodes.flatMap(node => List(node.left, node.right).filter(_ != null))
    }
    result
  }
}
